aspnetcore/mvc/advanced/custom-formatters.md

```diff
@@ -1,18 +1,101 @@
+
+## How to create a custom formatter class
+
+To create a formatter:
+
+* Derive the class from the appropriate base class.
+* Specify valid media types and encodings in the constructor.
+* Override `CanReadType`/`CanWriteType` methods
+* Override `ReadRequestBodyAsync`/`WriteResponseBodyAsync` methods
+  
+### Derive from the appropriate base class
+
+For text media types (for example, vCard), derive from the [TextInputFormatter](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.mvc.formatters.textinputformatter) or [TextOutputFormatter](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.mvc.formatters.textoutputformatter) base class.
+
+[!code-csharp[Main](custom-formatters/sample/Formatters/VcardOutputFormatter.cs?name=classdef)]
+
+For binary types, derive from the [InputFormatter](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.mvc.formatters.inputformatter) or [OutputFormatter](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.mvc.formatters.outputformatter) base class.
+
+### Specify valid media types and encodings
+
+In the constructor, specify valid media types and encodings by adding to the `SupportedMediaTypes` and `SupportedEncodings` collections.
+
+[!code-csharp[Main](custom-formatters/sample/Formatters/VcardOutputFormatter.cs?name=ctor&highlight=3,5-6)]
+
+> [!NOTE]  
+> You can't do constructor dependency injection in a formatter class. For example, you can't get a logger by adding a logger parameter to the constructor. To access services, you have to use the context object that gets passed in to your methods. A code example [below](#read-write) shows how to do this.
+
+### Override CanReadType/CanWriteType 
+
+Specify the type you can deserialize into or serialize from by overriding the `CanReadType` or `CanWriteType` methods. For example, you might only be able to create vCard text from a `Contact` type and vice versa.
+
+[!code-csharp[Main](custom-formatters/sample/Formatters/VcardOutputFormatter.cs?name=canwritetype)]
+
+For output formatters, in some scenarios you have to override `CanWriteResult` instead of `CanWriteType`. You need to use `CanWriteResult` if your action method returns a model class, and there are derived classes, and you need to know at runtime which derived class was returned by the action.  For example, suppose your action method signature returns a `Person` type, but it may return a `Student` or `Instructor` type that derives from `Person`. If you want your formatter to handle only `Student` objects, check the type of [Object](https://docs.microsoft.com/aspnet/core/api/microsoft.aspnetcore.mvc.formatters.outputformattercanwritecontext#Microsoft_AspNetCore_Mvc_Formatters_OutputFormatterCanWriteContext_Object) in the context object provided to the `CanWriteResult` method. Note that this only applies when the action method returns a model type; if the action method returns `IActionResult`, the `CanWriteType` method receives the runtime type.
```

Comment:
Not sure if this is any better (I have a list bias):
Use `CanWriteResult` when all these conditions are met in your action method:
- it returns a model class.
- the model can be a derived class.
- you need to know at runtime which derived class was returned. 

That sentence was hard for me to read. I don't think the machine translators we use will do very well.